src/rules/relayConnectionTypesSpecPinterestCustomization.ts

```diff
@@ -0,0 +1,113 @@
+import {
+  ASTVisitor,
+  ObjectTypeDefinitionNode,
+  TypeDefinitionNode,
+  ValidationContext,
+} from 'graphql';
+
+import { getNodeName, unwrapType, ValidationError } from '../utils';
+import { print } from 'graphql/language/printer';
+
+const MANDATORY_FIELDS = ['pageInfo', 'edges'];
+
+export function RelayConnectionTypesSpecPinterestCustomization(
+  context: ValidationContext,
+): ASTVisitor {
+  const ensureNameDoesNotEndWithConnection = (node: TypeDefinitionNode) => {
+    if (node.name.value.match(/Connection$/)) {
+      context.reportError(
+        new ValidationError(
+          'relay-connection-types-spec-pinterest-customization',
+          `Types that end in \`Connection\` must be an object type as per the relay spec. \`${node.name.value}\` is not an object type.`,
+          [node],
+        ),
+      );
+    }
+  };
+
+  return {
+    ScalarTypeDefinition: ensureNameDoesNotEndWithConnection,
+    InterfaceTypeDefinition: ensureNameDoesNotEndWithConnection,
+    UnionTypeDefinition: ensureNameDoesNotEndWithConnection,
+    EnumTypeDefinition: ensureNameDoesNotEndWithConnection,
+    InputObjectTypeDefinition: ensureNameDoesNotEndWithConnection,
+    ObjectTypeDefinition(node: ObjectTypeDefinitionNode) {
+      const typeName = node.name.value;
+      if (!typeName.endsWith('Connection')) {
+        return;
+      }
+      const fieldNames = node.fields?.map((field) => field.name.value);
+      const missingFields = MANDATORY_FIELDS.filter(
+        (requiredField) => fieldNames?.indexOf(requiredField) === -1,
+      );
+
+      if (missingFields.length) {
+        context.reportError(
+          new ValidationError(
+            'relay-connection-types-spec-pinterest-customization',
+            `Connection \`${typeName}\` is missing the following field${
+              missingFields.length > 1 ? 's' : ''
+            }: ${missingFields.join(', ')}.`,
+            [node],
+          ),
+        );
+        return;
+      }
+
+      // Validates that the edges field type is of a valid type
```

Comment:
Perhaps a better way to structure this could be:

```
function validateConnection(node: ...): {
     // checks that a node that ends with Connection has pageInfo and edges
}

validateConnection(connectionNode)
edges, pageInfo = getFields(connectionNode)

validateEdges(edges)
edge = getEdge(edges)
validateEdge(edge)

validate(pageInfo)
```

Reply:
Thanks, will restructure this.

The only thing that will be left out from this is this part:

```
edge = getEdge(edges)
validateEdge(edge)
```

This part is validated by `relay-edge-types-spec-pinterest-customization`.